### PR TITLE
Improve performance of applyDeletedShards

### PR DESCRIPTION
**Pull request by jasontedor**
This commit addresses a performance issue in
IndicesClusterStateService#applyDeletedShards. Namely, the current
implementation is O(number of indices \* number of shards). This is
because of an outer loop over the indices and an inner loop over the
assigned shards, all to check if a shard is in the outer index. Instead,
we can group the shards by index, and then just do a map lookup for each
index.

Testing this on a single-node with 2500 indices, each with 2 shards,
creating an index before this optimization takes 0.90s and after this
optimization takes 0.19s.

Relates #18776

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java [hunk 1]
```diff
@@ -285,19 +288,23 @@ private void applyDeletedShards(final ClusterChangedEvent event) {
         if (routingNode == null) {
             return;
         }
-        Set<String> newShardAllocationIds = new HashSet<>();
+
+        final Map<Index, List<ShardRouting>> shardsByIndex = new HashMap<>();
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: Why not go directly to Map<Index, Set<String>>? That we only need to build each set one time?

**Comment by nik9000** [comment]
Left a minor but LGTM.

**Comment by s1monw** [comment]
LGTM

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: Good suggestion. I pushed adc635c1b200f6213cd5c1db18e028e94c4578de.

**Comment by nik9000** [comment]
LGTM too.

**Comment by ywelsch** [comment]
This is already addressed by #17270 (https://github.com/elastic/elasticsearch/pull/17270/files/f71f7de3e85191ae0ac31191d2b6f1ba173c73bd#diff-e3d064cfd24df0c1f53af98e6f9bc66dR372) which also does not even require building an extra map :-) Interestingly, both reviewed by @s1monw ;-) This one is still useful, however, if we want to backport to 2.4.x.

**Comment by s1monw** [comment]
>  both reviewed by @s1monw ;-) 

this one had a different focus :)

**Pull request merged**